File: src/ipc.rs
use base64::Engine;
use std::fmt::Debug;

#[repr(u8)]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum MessageType {
    Evaluate = 0,
    Respond = 1,
}

/// Binary protocol message types.
///
/// The binary format uses aligned buffers for efficient memory access:
/// - First 12 bytes: three u32 offsets (u16_offset, u8_offset, str_offset)
/// - u32 buffer: from byte 12 to u16_offset
/// - u16 buffer: from u16_offset to u8_offset
/// - u8 buffer: from u8_offset to str_offset
/// - string buffer: from str_offset to end
///
/// Message format in the u8 buffer:
/// - First u8: message type (0 = Evaluate, 1 = Respond)
/// - Remaining data depends on message type
///
/// Evaluate format (supports batching - multiple operations in one message):
/// - u8: message type (0)
/// - For each operation (read until buffer exhausted):
///   - u32: function ID
///   - encoded arguments (varies by function)
///
/// Respond format:
/// - u8: message type (1)
/// - For each operation result:
///   - encoded return value (varies by function)
#[derive(Debug)]
pub(crate) struct IPCMessage {
    data: Vec<u8>,
}

impl IPCMessage {
    pub fn new(data: Vec<u8>) -> Self {
        Self { data }
    }

    pub fn new_respond(push_data: impl FnOnce(&mut EncodedData)) -> Self {
        let mut encoder = EncodedData::new();
        encoder.push_u8(MessageType::Respond as u8);

        push_data(&mut encoder);

        IPCMessage::new(encoder.to_bytes())
    }

    pub fn ty(&self) -> Result<MessageType, ()> {
        let mut decoded = DecodedData::from_bytes(&self.data)?;
        let message_type = decoded.take_u8()?;
        match message_type {
            0 => Ok(MessageType::Evaluate),
            1 => Ok(MessageType::Respond),
            _ => Err(()),
        }
    }

    pub fn decoded(&self) -> Result<DecodedVariant<'_>, ()> {
        let mut decoded = DecodedData::from_bytes(&self.data)?;
        let message_type = decoded.take_u8()?;
        let message_type = match message_type {
            0 => DecodedVariant::Evaluate { data: decoded },
            1 => DecodedVariant::Respond { data: decoded },
            _ => return Err(()),
        };
        Ok(message_type)
    }

    pub fn data(&self) -> &[u8] {
        &self.data
    }

    pub fn into_data(self) -> Vec<u8> {
        self.data
    }
}

pub enum DecodedVariant<'a> {
    Respond { data: DecodedData<'a> },
    Evaluate { data: DecodedData<'a> },
}

/// Decoded binary data with aligned buffer access
pub(crate) struct DecodedData<'a> {
    u8_buf: &'a [u8],
    u16_buf: &'a [u16],
    u32_buf: &'a [u32],
    str_buf: &'a [u8],
}

impl<'a> DecodedData<'a> {
    pub fn from_bytes(bytes: &'a [u8]) -> Result<Self, ()> {
        if bytes.len() < 12 {
            return Err(());
        }

        let header: [u32; 3] = bytemuck::cast_slice(&bytes[0..12])
            .try_into()
            .map_err(|_| ())?;
        let [u16_offset, u8_offset, str_offset] = header;

        let u32_buf = bytemuck::cast_slice(&bytes[12..u16_offset as usize]);
        let u16_buf = bytemuck::cast_slice(&bytes[u16_offset as usize..u8_offset as usize]);
        let u8_buf = &bytes[u8_offset as usize..str_offset as usize];
        let str_buf = &bytes[str_offset as usize..];

        Ok(Self {
            u8_buf,
            u16_buf,
            u32_buf,
            str_buf,
        })
    }

    pub fn take_u8(&mut self) -> Result<u8, ()> {
        let [first, rest @ ..] = &self.u8_buf else {
            return Err(());
        };
        self.u8_buf = rest;
        Ok(*first)
    }

    pub fn take_u16(&mut self) -> Result<u16, ()> {
        let [first, rest @ ..] = &self.u16_buf else {
            return Err(());
        };
        self.u16_buf = rest;
        Ok(*first)
    }

    pub fn take_u32(&mut self) -> Result<u32, ()> {
        let [first, rest @ ..] = &self.u32_buf else {
            return Err(());
        };
        self.u32_buf = rest;
        Ok(*first)
    }

    pub fn take_u64(&mut self) -> Result<u64, ()> {
        let low = self.take_u32()? as u64;
        let high = self.take_u32()? as u64;
        Ok((high << 32) | low)
    }

    pub fn take_str(&mut self) -> Result<&'a str, ()> {
        let len = self.take_u32()? as usize;
        let Some((buf, rem)) = self.str_buf.split_at_checked(len) else {
            return Err(());
        };
        let s = std::str::from_utf8(buf).map_err(|_| ())?;
        self.str_buf = rem;
        Ok(s)
    }
}

/// Encoder for building binary messages
#[derive(Debug, Default)]
pub(crate) struct EncodedData {
    u8_buf: Vec<u8>,
    u16_buf: Vec<u16>,
    u32_buf: Vec<u32>,
    str_buf: Vec<u8>,
}

impl EncodedData {
    pub fn new() -> Self {
        Self {
            u8_buf: Vec::new(),
            u16_buf: Vec::new(),
            u32_buf: Vec::new(),
            str_buf: Vec::new(),
        }
    }

    pub fn byte_len(&self) -> usize {
        12 + self.u32_buf.len() * 4 + self.u16_buf.len() * 2 + self.u8_buf.len() + self.str_buf.len()
    }

    pub fn push_u8(&mut self, value: u8) {
        self.u8_buf.push(value);
    }

    pub fn push_u16(&mut self, value: u16) {
        self.u16_buf.push(value);
    }

    pub fn push_u32(&mut self, value: u32) {
        self.u32_buf.push(value);
    }

    pub fn push_u64(&mut self, value: u64) {
        self.push_u32((value & 0xFFFFFFFF) as u32);
        self.push_u32((value >> 32) as u32);
    }

    pub fn push_str(&mut self, value: &str) {
        self.push_u32(value.len() as u32);
        self.str_buf.extend_from_slice(value.as_bytes());
    }

    pub fn to_bytes(&self) -> Vec<u8> {
        let u16_offset = 12 + self.u32_buf.len() * 4;
        let u8_offset = u16_offset + self.u16_buf.len() * 2;
        let str_offset = u8_offset + self.u8_buf.len();

        let total_len = str_offset + self.str_buf.len();
        let mut bytes = Vec::with_capacity(total_len);

        // Write header offsets
        bytes.extend_from_slice(&(u16_offset as u32).to_le_bytes());
        bytes.extend_from_slice(&(u8_offset as u32).to_le_bytes());
        bytes.extend_from_slice(&(str_offset as u32).to_le_bytes());

        // Write u32 buffer
        for &u in &self.u32_buf {
            bytes.extend_from_slice(&u.to_le_bytes());
        }

        // Write u16 buffer
        for &u in &self.u16_buf {
            bytes.extend_from_slice(&u.to_le_bytes());
        }

        // Write u8 buffer
        bytes.extend_from_slice(&self.u8_buf);

        // Write string buffer
        bytes.extend_from_slice(&self.str_buf);

        bytes
    }

    pub fn extend(&mut self, other: &EncodedData) {
        self.u8_buf.extend_from_slice(&other.u8_buf);
        self.u16_buf.extend_from_slice(&other.u16_buf);
        self.u32_buf.extend_from_slice(&other.u32_buf);
        self.str_buf.extend_from_slice(&other.str_buf);
    }
}

pub(crate) fn decode_data(bytes: &[u8]) -> Option<IPCMessage> {
    // Decode base64 header
    let engine = base64::engine::general_purpose::STANDARD;
    let data = engine.decode(bytes).ok()?;

    Some(IPCMessage { data })
}

File: src/ts/ipc.ts
/**
 * Binary Protocol Encoder/Decoder
 *
 * The binary format uses aligned buffers for efficient memory access:
 * - First 12 bytes: three u32 offsets (u16_offset, u8_offset, str_offset)
 * - u32 buffer: from byte 12 to u16_offset
 * - u16 buffer: from u16_offset to u8_offset
 * - u8 buffer: from u8_offset to str_offset
 * - string buffer: from str_offset to end
 *
 * Message format in the u8 buffer:
 * - First u8: message type (0 = Evaluate, 1 = Respond)
 * - Remaining data depends on message type
 */

import { DataDecoder, DataEncoder } from "./encoding";
import { getFunctionRegistry, getTypeCache, CachedTypeInfo } from "./function_registry";
import { parseTypeDef, TypeClass } from "./types";

enum MessageType {
  Evaluate = 0,
  Respond = 1,
}

// Type caching markers - must match Rust's TYPE_CACHED and TYPE_FULL
const TYPE_CACHED = 0xff;
const TYPE_FULL = 0xfe;

// Reserved function ID for dropping heap refs - must match Rust's DROP_HEAP_REF_FN_ID
const DROP_HEAP_REF_FN_ID = 0xffffffff;

// Reserved function ID for cloning heap refs - must match Rust's CLONE_HEAP_REF_FN_ID
const CLONE_HEAP_REF_FN_ID = 0xfffffffe;

// Reserved function ID for dropping native Rust refs - must match Rust's DROP_NATIVE_REF_FN_ID
const DROP_NATIVE_REF_FN_ID = 0xffffffff;

/**
 * Sends binary data to Rust and receives binary response.
 */
function sync_request_binary(
  endpoint: string,
  data: ArrayBuffer
): ArrayBuffer | null {
  const xhr = new XMLHttpRequest();
  xhr.open("POST", endpoint, false);
  xhr.responseType = "arraybuffer";

  // Encode as base64 for header (Android workaround)
  const bytes = new Uint8Array(data);
  let binary = "";
  for (let i = 0; i < bytes.length; i++) {
    binary += String.fromCharCode(bytes[i]);
  }
  const base64 = btoa(binary);
  xhr.setRequestHeader("dioxus-data", base64);
  xhr.send();

  if (xhr.status === 200 && xhr.response && xhr.response.byteLength > 0) {
    return xhr.response as ArrayBuffer;
  }
  return null;
}

/**
 * Entry point for Rust to call JS functions using binary protocol.
 * Handles batched operations - reads and executes operations until buffer is exhausted.
 *
 * @param dataBase64 - Base64 encoded binary data containing message with operations
 */
function evaluate_from_rust_binary(dataBase64: string): DataDecoder | null {
  // Decode base64 to ArrayBuffer
  const binary = atob(dataBase64);
  const bytes = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) {
    bytes[i] = binary.charCodeAt(i);
  }
  return handleBinaryResponse(bytes.buffer);
}

/**
 * Parse type information from the decoder.
 * Handles both cached and full type definitions.
 */
function parseTypeInfo(decoder: DataDecoder): CachedTypeInfo {
  const typeCache = getTypeCache();
  const typeMarker = decoder.takeU8();

  if (typeMarker === TYPE_CACHED) {
    // Cached type - look up by ID
    const typeId = decoder.takeU32();
    const cached = typeCache.get(typeId);
    if (!cached) {
      throw new Error(`Unknown cached type ID: ${typeId}`);
    }
    return cached;
  } else if (typeMarker === TYPE_FULL) {
    // Full type definition - parse and cache
    const typeId = decoder.takeU32();
    const paramCount = decoder.takeU8();

    // Get the remaining bytes for parsing type definitions
    const typeBytes = decoder.getRemainingBytes();
    const offset = { value: 0 };

    const paramTypes: TypeClass[] = [];
    for (let i = 0; i < paramCount; i++) {
      paramTypes.push(parseTypeDef(typeBytes, offset));
    }
    const returnType = parseTypeDef(typeBytes, offset);

    // Advance the decoder past the type definition bytes we consumed
    decoder.skipBytes(offset.value);

    const cached: CachedTypeInfo = { paramTypes, returnType };
    typeCache.set(typeId, cached);
    return cached;
  } else {
    throw new Error(`Unknown type marker: ${typeMarker}`);
  }
}

/**
 * Handle binary response from Rust.
 * May contain nested Evaluate calls (for callbacks).
 */
function handleBinaryResponse(
  response: ArrayBuffer | null
): DataDecoder | null {
  if (!response || response.byteLength === 0) {
    return null;
  }

  const decoder = new DataDecoder(response);
  const rawMsgType = decoder.takeU8();
  const msgType: MessageType = rawMsgType;

  if (msgType === MessageType.Respond) {
    // Respond - just return the decoder for further processing
    return decoder;
  } else if (msgType === MessageType.Evaluate) {
    // Evaluate - Rust is calling JS functions (possibly multiple)

    const encoder = new DataEncoder();
    encoder.pushU8(MessageType.Respond);

    // Process all operations
    while (decoder.hasMoreU32()) {
      const fnId = decoder.takeU32();
      // Parse type information (cached or full)
      const typeInfo = parseTypeInfo(decoder);

      // Handle special drop function
      if (fnId === DROP_HEAP_REF_FN_ID) {
        const heapId = decoder.takeU64();
        window.jsHeap.remove(heapId);
        continue;
      }

      // Handle special clone function
      if (fnId === CLONE_HEAP_REF_FN_ID) {
        const heapId = decoder.takeU64();
        const value = window.jsHeap.get(heapId);
        const newId = window.jsHeap.insert(value);
        encoder.pushU64(newId);
        continue;
      }

      // Get the raw JS function
      const functionRegistry = getFunctionRegistry();
      const jsFunction = functionRegistry[fnId];
      if (!jsFunction) {
        throw new Error("Unknown function ID in response: " + fnId);
      }

      // Decode parameters using their respective types
      const params = typeInfo.paramTypes.map((paramType) => paramType.decode(decoder));
      console.log("[IPC] Calling JS function ID:", fnId, "with params:", params);

      // Call the original JS function with decoded parameters
      const result = jsFunction(...params);

      console.log("[IPC] JS function ID:", fnId, "returned:", result);

      // Encode the result using the return type
      typeInfo.returnType.encode(encoder, result);
    }

    const nextResponse = sync_request_binary(
      "wry://handler",
      encoder.finalize()
    );
    return handleBinaryResponse(nextResponse);
  }

  return null;
}

export {
  evaluate_from_rust_binary,
  handleBinaryResponse,
  sync_request_binary,
  MessageType,
  DROP_NATIVE_REF_FN_ID,
};
